### PR TITLE
Gens better doc handling for what is used for generation

### PR DESCRIPTION
**Pull request by Uzay-G**
- implements better mechanisms for deciding what goes into generation
- adds a token count to datasource blocks

### CODE REVIEW

File: core/src/data_sources/data_source.rs [hunk 1]
```diff
@@ -805,6 +808,8 @@ impl DataSource {
                         let collection = self.qdrant_collection();
                         let chunk_size = self.config.max_chunk_size;
                         let qdrant_client = l_qdrant_client.clone();
+                        let mut token_count = chunks.len() * chunk_size;
+                        d.token_count = Some(token_count);
```
Comment (spolu) [inline core/src/data_sources/data_source.rs:812]: Do we need that line??

File: front/pages/w/[wId]/u/gens/index.tsx [hunk 2]
```diff
@@ -807,25 +807,36 @@ export default function AppGens({
...
+      if (d.tokenCount <= space) {
+        extracts.push({
+          documentId: d.documentId,
+          extract: d.text,
+          score: d.score.toFixed(2),
+        });
+        space -= d.tokenCount;
+      }
+      return space;
+    }, 8000);
```
Comment (spolu) [inline front/pages/w/[wId]/u/gens/index.tsx:839]: Maybe a bit less? if we account for prompt + noise around chunks we may overflow here no?

File: front/pages/w/[wId]/u/gens/index.tsx [hunk 3]
```diff
@@ -925,6 +936,7 @@ export default function AppGens({
       DustProdActionRegistry["gens-retrieval"].config
     );
 
+    config.DATASOURCE.target_document_tokens = 800;
```
Comment (spolu) [inline front/pages/w/[wId]/u/gens/index.tsx:939]: The goal is to get 3 chunks here right?

File: front/pages/w/[wId]/u/gens/index.tsx [hunk 4]
```diff
@@ -955,6 +967,7 @@ export default function AppGens({
       if (event.type === "block_execution") {
         const e = event.content.execution[0][0];
         if (event.content.block_name === "OUTPUT") {
+          console.log(e.value.retrievals);
```
Comment (spolu) [inline front/pages/w/[wId]/u/gens/index.tsx:970]: clean up?

File: front/pages/w/[wId]/u/gens/index.tsx (hunk 3 above)
Comment (Uzay-G) [reply front/pages/w/[wId]/u/gens/index.tsx:939]: Yup, we could adjust this

File: core/src/data_sources/data_source.rs (hunk 1 above)
Comment (Uzay-G) [reply core/src/data_sources/data_source.rs:812]: Yes, the expansion is interrupted early if there isn't enough space/chunks, so this handles that case

File: front/pages/w/[wId]/u/gens/index.tsx (hunk 2 above)
Comment (Uzay-G) [reply front/pages/w/[wId]/u/gens/index.tsx:839]: 7000?

Comment (spolu) [reply front/pages/w/[wId]/u/gens/index.tsx]: Sure let's do multiple of 1024 so 7168

File: core/src/data_sources/data_source.rs (hunk 1 above)
Comment (spolu) [reply core/src/data_sources/data_source.rs:812]: ack :+1:

File: front/pages/w/[wId]/u/gens/index.tsx [hunk 5]
```diff
@@ -807,26 +807,36 @@ export default function AppGens({
...
+      if (d.tokenCount <= space) {
+        extracts.push({
+          documentId: d.documentId,
+          extract: d.text,
+          score: d.score.toFixed(2),
+        });
+        space -= d.tokenCount;
+      }
+      return space;
+    }, 7000);
```
Comment (spolu) [inline front/pages/w/[wId]/u/gens/index.tsx:839]: Let's do 7168

File: front/pages/w/[wId]/u/gens/index.tsx (hunk 3 above)
Comment (spolu) [reply front/pages/w/[wId]/u/gens/index.tsx:935]: Let's do 256*3

**Review by spolu — DISMISSED** [review]
Final nits but feel free to merge after that :+1:

**Comment by spolu** [comment]
(happy to deploy core once this is merged :+1: )

**Pull request merged**